packages/components/src/validated-form-controls/components/stories/overview.story.tsx

```diff
@@ -73,28 +96,113 @@ export const WithMultipleControls: Story = {
 export const WithHelpTextReplacement: Story = {
 	render: function Template() {
 		const [ text, setText ] = useState( '' );
-		const [ hasCustomError, setHasCustomError ] = useState( false );
+		const [ customValidityMessage, setCustomValidityMessage ] =
+			useState<
+				React.ComponentProps<
+					typeof ValidatedInputControl
+				>[ 'customValidityMessage' ]
+			>( undefined );
 
 		return (
 			<ValidatedInputControl
 				label="Text"
 				required
 				value={ text }
 				help={
-					hasCustomError
+					customValidityMessage
 						? undefined
 						: 'The word "error" is not allowed.'
 				}
-				customValidator={ ( value ) => {
+				onValidate={ ( value ) => {
 					if ( value?.toLowerCase() === 'error' ) {
-						setHasCustomError( true );
-						return 'The word "error" is not allowed.';
+						setCustomValidityMessage( {
+							type: 'invalid',
+							message: 'The word "error" is not allowed.',
+						} );
+					} else {
+						setCustomValidityMessage( undefined );
 					}
-					setHasCustomError( false );
-					return undefined;
 				} }
 				onChange={ ( value ) => setText( value ?? '' ) }
+				customValidityMessage={ customValidityMessage }
 			/>
 		);
 	},
 };
+
+/**
+ * To provide feedback from server-side validation, the `customValidityMessage` prop can be used
+ * to show additional status indicators while waiting for the server response,
+ * and after the response is received.
+ *
+ * These indicators are intended for asynchronous validation calls that may take more than 1 second to complete.
+ * They may be unnecessary when responses are generally quick.
+ */
+export const AsyncValidation: StoryObj< typeof ValidatedInputControl > = {
```

Comment:
@oandregal I don't know how much we want to abstract this in the DataForm APIs, but at the component level I tried to keep it flexible for now. If real-world use cases start to show patterns, maybe we can abstract it a bit more at the component level as well.

The biggest unknowns for me right now are:

- What percentage of use cases will need async?
- What percentage of those are slow enough to need a `validating` state?